### PR TITLE
Ignore prometheus metrics when their values are NaN or Inf

### PR DESCRIPTION
**Pull request by kaiyan-sheng**
When prometheus report metrics with value NaN or +Inf or -Inf, metricbeat will fail with error like "Failed to serialize the event: unsupported float value: NaN". This PR is to add the logic to ignore prometheus metrics with NaN/Inf value in `collector` metricset.

This PR also added a `metrics-with-naninf.plain` test data with NaN/Inf as metric value. From the output `metrics-with-naninf.plain-expected.json` you can see, the metrics with NaN/Inf value are not there. 

closes https://github.com/elastic/beats/issues/10849

### CODE REVIEW

File: metricbeat/module/prometheus/collector/_meta/testdata/histogram.plain-expected.json [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        },
+        "metricset": {
+            "name": "collector"
+        },
+        "prometheus": {
+            "labels": {
+                "code": "200"
+            },
+            "metrics": {
+                "hash_seconds_count": 0,
```
Comment (kaiyan-sheng) [inline metricbeat/module/prometheus/collector/_meta/testdata/histogram.plain-expected.json]: Can anyone help me understand why this metric exists please? I checked regenerating this json file without the change to ignore NaN, this metric still got generated. Is this something related to histogram?

**Comment by kaiyan-sheng** [comment]
Changelog will be added later.

File: metricbeat/module/prometheus/collector/collector.go [hunk 2]
```diff
@@ -29,6 +28,7 @@ import (
 const (
 	defaultScheme = "http"
 	defaultPath   = "/metrics"
+	nanInf        = 9223372036854775808
```
Comment (exekias) [inline metricbeat/module/prometheus/collector/collector.go]: where did you find this constant? Isn't it public in the Prometheus client?

File: metricbeat/module/prometheus/collector/_meta/testdata/histogram.plain-expected.json (hunk 1 above)
Comment (exekias) [reply metricbeat/module/prometheus/collector/_meta/testdata/histogram.plain-expected.json]: yes, check https://github.com/elastic/beats/blob/master/metricbeat/module/prometheus/collector/data.go#L106, I guess numbers are 0 here because of NaN

Comment (kaiyan-sheng) [reply metricbeat/module/prometheus/collector/_meta/testdata/histogram.plain-expected.json]: @exekias ahh it's the `sum` and `count` from histogram metrics. I did not specify a `sum` or `count` here in `histogram.plain` and that's why its showing 0.

File: metricbeat/module/prometheus/collector/collector.go (hunk 2 above)
Comment (kaiyan-sheng) [reply metricbeat/module/prometheus/collector/collector.go]: Oops I used the hardcoded value instead... It should be something like `math.NaN()`. I will make the change! Thanks!

**Comment by kaiyan-sheng** [comment]
> I'm guessing Prometheus helper is also affected: https://github.com/elastic/beats/blob/master/metricbeat/helper/prometheus/prometheus.go

@exekias Yes you are right, I just pushed a new commit with changes in the prometheus helper. Thanks!

File: metricbeat/helper/prometheus/prometheus_test.go [hunk 3]
```diff
@@ -98,6 +98,22 @@ metrics_one_midichlorians_bucket{rank="padawan",alive="yes",le="+Inf"} 28
 metrics_one_midichlorians_sum{rank="padawan",alive="yes"} 800001
 metrics_one_midichlorians_count{rank="padawan",alive="yes"} 28
 
+`
+
+	promHistogramKeyLabelWithNaNInf = `
+# TYPE metrics_one_midichlorians histogram
+metrics_one_midichlorians_bucket{rank="youngling",alive="yes",le="100"} NaN
+metrics_one_midichlorians_bucket{rank="youngling",alive="yes",le="500"} +Inf
+metrics_one_midichlorians_bucket{rank="youngling",alive="yes",le="1000"} -Inf
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus_test.go]: It would be really really weird to find these values here, because they are integers, and integers cannot have these values in Go. Also this is a histogram, these values here would be incorrect. In this case this could be considered as invalid as finding an string.

The issue observed in #10849 was with gauges, something like:
```
# HELP kafka_consumer_records_lag_records The latest lag of the partition
# TYPE kafka_consumer_records_lag_records gauge
kafka_consumer_records_lag_records{client_id="some-consumer",} NaN
kafka_consumer_records_lag_records{client_id="other-consumer",} NaN
kafka_consumer_records_lag_records{client_id="another-consumer",} NaN
```

Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/prometheus_test.go]: Yeah I agree, this should not happen on histogram. I will remove the code as well as the test!

**Comment by kaiyan-sheng** [comment]
@jsoriano I don't know how I got myself into this histogram _hole_ and ignored all the other real problem 🤣 Thanks for the explanation (and digging me out of the histogram  _hole_ )! I added the part for Gauge/Counter/Summary and removed the histogram change. Please let me know if this PR is going the right direction now... 😃

File: metricbeat/helper/prometheus/prometheus.go [hunk 4]
```diff
@@ -165,6 +166,14 @@ func (p *prometheus) GetProcessedMetrics(mapping *MetricsMapping) ([]common.MapS
 
 				// value may be a mapstr (for histograms and summaries), do a deep update to avoid smashing existing fields
 				update := common.MapStr{}
+				if valueFloat64, ok := value.(float64); ok {
+					if math.IsNaN(valueFloat64) || math.IsInf(valueFloat64, 1) || math.IsInf(valueFloat64, 0) {
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus.go]: `math.IsInf(value, 0)` checks if the value is `+Inf`  or `-Inf`.
```suggestion
					if math.IsNaN(valueFloat64) || math.IsInf(valueFloat64, 0) {
```

File: metricbeat/helper/prometheus/prometheus.go [hunk 5]
```diff
@@ -165,6 +166,14 @@ func (p *prometheus) GetProcessedMetrics(mapping *MetricsMapping) ([]common.MapS
 
 				// value may be a mapstr (for histograms and summaries), do a deep update to avoid smashing existing fields
 				update := common.MapStr{}
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus.go]: Move this line and the comment above after the added code where is the related code.

File: metricbeat/helper/prometheus/prometheus.go [hunk 6]
```diff
@@ -165,6 +166,14 @@ func (p *prometheus) GetProcessedMetrics(mapping *MetricsMapping) ([]common.MapS
...
+				if valueFloat64, ok := value.(float64); ok {
+					if math.IsNaN(valueFloat64) || math.IsInf(valueFloat64, 1) || math.IsInf(valueFloat64, 0) {
+						// Remove the event if value is NaN or +Inf or -Inf
+						delete(eventsMap, keyLabels.String())
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus.go]: I think this will remove the whole event for metrics having the same labels, I don't think we want that. If you want to remove the event that could have been created on `getEvent` for this metric, you could instead move `getEvent` after this check, so it is only called if the value is going to be stored.

File: metricbeat/helper/prometheus/prometheus_test.go [hunk 7]
```diff
@@ -69,6 +69,15 @@ metrics_one_count_total{name="jane",surname="foster"} 1
 metrics_one_count_total{name="john",surname="williams"} 2
 metrics_one_count_total{name="jahn",surname="baldwin",age="30"} 3
 
+`
+
+	promGaugeKeyLabelWithNaNInf = `
+# TYPE metrics_one_count_total gauge
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus_test.go]: Maybe add here some case of metric with the same labels than the discarded ones, to check that metrics with the same labels are still being collected.
```suggestion
# TYPE metrics_one_count_errors gauge
metrics_one_count_errors{name="jane",surname="foster"} 0
# TYPE metrics_one_count_total gauge
```

File: metricbeat/module/prometheus/collector/data.go [hunk 8]
```diff
@@ -55,33 +55,39 @@ func getPromEventsFromMetricFamily(mf *dto.MetricFamily) []PromEvent {
 
 		counter := metric.GetCounter()
 		if counter != nil {
-			events = append(events, PromEvent{
-				data: common.MapStr{
-					name: counter.GetValue(),
-				},
-				labels: labels,
-			})
+			if !math.IsNaN(counter.GetValue()) && !math.IsInf(counter.GetValue(), 1) && !math.IsInf(counter.GetValue(), 0) {
```
Comment (jsoriano) [inline metricbeat/module/prometheus/collector/data.go]: Also here, checking `IsNan(v)` and `IsInf(v, 0)` would be enough.

File: metricbeat/module/prometheus/collector/_meta/testdata/gauge-with-naninf.plain [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# HELP kafka_consumer_records_lag_records The latest lag of the partition
+# TYPE kafka_consumer_records_lag_records gauge
+kafka_consumer_records_lag_records{client_id="consumer1",} NaN
+kafka_consumer_records_lag_records{client_id="consumer2",} +Inf
+kafka_consumer_records_lag_records{client_id="consumer3",} -Inf
+kafka_consumer_records_lag_records{client_id="consumer4",} 5
+
```
Comment (jsoriano) [inline metricbeat/module/prometheus/collector/_meta/testdata/gauge-with-naninf.plain]: Could you add here also some examples of the other metric types?

File: metricbeat/helper/prometheus/prometheus.go (hunk 4 above)
Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/prometheus.go]: ohh yes you are right, thanks!

File: metricbeat/helper/prometheus/prometheus.go (hunk 5 above)
Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/prometheus.go]: done!

File: metricbeat/helper/prometheus/prometheus.go (hunk 6 above)
Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/prometheus.go]: Yes you are right! I did not want to remove all events with the same labels. Thanks for catching this!

**Comment by kaiyan-sheng** [comment]
I'm not done with addressing Jaime's comments yet 😬 Will continue tomorrow!

File: metricbeat/module/prometheus/collector/_meta/testdata/gauge-with-naninf.plain (hunk 9 above)
Comment (kaiyan-sheng) [reply metricbeat/module/prometheus/collector/_meta/testdata/gauge-with-naninf.plain]: yes, I will put them in the same file and rename the file 👍

File: metricbeat/helper/prometheus/metric.go [hunk 10]
```diff
@@ -135,28 +135,33 @@ func (m *commonMetric) GetField() string {
...
 
 	summary := metric.GetSummary()
 	if summary != nil {
 		value := common.MapStr{}
-		value["sum"] = summary.GetSampleSum()
-		value["count"] = summary.GetSampleCount()
+		if !math.IsNaN(summary.GetSampleSum()) && !math.IsInf(summary.GetSampleSum(), 0) {
+			value["sum"] = summary.GetSampleSum()
+			value["count"] = summary.GetSampleCount()
+		}
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/metric.go]: Should we also do this same check in the histogram sum?

File: metricbeat/helper/prometheus/metric.go [hunk 11]
```diff
@@ -135,28 +135,33 @@ func (m *commonMetric) GetField() string {
 func (m *commonMetric) GetValue(metric *dto.Metric) interface{} {
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/metric.go]: We could do some refactors to split this method in smaller methods and to reuse the check code, but nothing to do in this PR.

File: metricbeat/helper/prometheus/metric.go (hunk 10 above)
Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/metric.go]: haha I removed it because histogram `bucket` value are `uint64` type instead of `float64` so it will not be NaN or Inf. But instead, it will be `9223372036854775808` 😄 I added the check back into histogram so  we won't get huge numbers like this 👍

Comment (jsoriano) [reply metricbeat/helper/prometheus/metric.go]: Well, I said it mainly for the `sum`, that is a `float64`, but I see you added the code for all the values :+1:

Comment (kaiyan-sheng) [reply metricbeat/helper/prometheus/metric.go]: yeah `sum` is float64 and the other values are uint 😬

File: metricbeat/module/prometheus/collector/data.go [hunk 12]
```diff
@@ -55,36 +55,42 @@ func getPromEventsFromMetricFamily(mf *dto.MetricFamily) []PromEvent {
...
 
 		gauge := metric.GetGauge()
 		if gauge != nil {
-			events = append(events, PromEvent{
-				data: common.MapStr{
-					name: gauge.GetValue(),
-				},
-				labels: labels,
-			})
+			if !math.IsNaN(gauge.GetValue()) && !math.IsInf(gauge.GetValue(), 1) && !math.IsInf(gauge.GetValue(), 0) {
```
Comment (jsoriano) [inline metricbeat/module/prometheus/collector/data.go]: Oops, I didn't see it.
```suggestion
			if !math.IsNaN(gauge.GetValue()) && !math.IsInf(gauge.GetValue(), 0) {
```

Comment (kaiyan-sheng) [reply metricbeat/module/prometheus/collector/data.go]: Oops! My fault! Thanks for catching it!!

**Comment by yxyx520** [comment]
with the version v7.2.0, I still get an error "2019-07-17T09:26:44.695Z	ERROR	elasticsearch/client.go:394	Failed to encode event: unsupported float value: NaN"

**Comment by yxyx520** [comment]
@kaiyan-sheng any suggestions ?

**Comment by charkadiusz** [comment]
@kaiyan-sheng, same here - v 7.3.0 and am getting 
```
ERROR	elasticsearch/client.go:402	Failed to encode event: unsupported float value: NaN
```
 in the prometheus collector metricset when scraping `/federate` endpoint on prometheus in a kubernetes cluster.
This is quite problematic as I'm getting flooded with those error logs. Any advice?
